phoenix-core/src/it/java/org/apache/phoenix/end2end/MigrateSystemTablesToSystemNamespaceIT.java

```diff
@@ -0,0 +1,399 @@
+    }
+
+    // Tests that OLDER clients fail after system tables are migrated
+    // Clients should be restarted with new properties which are consistent on both client and server
+    @Test
+    public void oldClientsAfterSystemTableMigrationShouldFail() throws IOException, InterruptedException {
+
+        user1.doAs(new PrivilegedExceptionAction<Void>() {
+            @Override
+            public Void run() throws Exception {
+                createConnection(getClientPropertiesWithSystemMappingEnabled());
+                return null;
+            }
+        });
+
+        hbaseTables = getHBaseTables();
+        assertTrue(hbaseTables.size() == PHOENIX_NAMESPACE_MAPPED_SYSTEM_TABLES.size());
+        assertTrue(hbaseTables.containsAll(PHOENIX_NAMESPACE_MAPPED_SYSTEM_TABLES));
+
+        try {
+            user2.doAs(new PrivilegedExceptionAction<Void>() {
+                @Override
+                public Void run() throws Exception {
+                    createConnection(getClientPropertiesWithSystemMappingDisabled());
+                    return null;
+                }
+            });
+            fail("Client should not be able to connect to cluster with inconsistent SYSTEM table namespace properties");
+        } catch (Exception e) {
+            //ignore
+        }
+
+        hbaseTables = getHBaseTables();
+        assertTrue(hbaseTables.size() == PHOENIX_NAMESPACE_MAPPED_SYSTEM_TABLES.size());
+        assertTrue(hbaseTables.containsAll(PHOENIX_NAMESPACE_MAPPED_SYSTEM_TABLES));
+    }
+
+    // Tests that only one client can migrate the system table to system namespace
+    // Migrate process acquires lock in SYSMUTEX table
+    @Test
+    public void onlyOneClientCanMigrate() throws IOException, InterruptedException, SQLException {
+
+        user1.doAs(new PrivilegedExceptionAction<Void>() {
+            @Override
+            public Void run() throws Exception {
+                createConnection(getClientPropertiesWithSystemMappingDisabled());
+                return null;
+            }
+        });
+
+        hbaseTables = getHBaseTables();
+        assertTrue(hbaseTables.size() == PHOENIX_SYSTEM_TABLES.size());
+        assertTrue(hbaseTables.containsAll(PHOENIX_SYSTEM_TABLES));
+
+        user2.doAs(new PrivilegedExceptionAction<Void>() {
+            @Override
+            public Void run() throws Exception {
+                // Acquire Mutex Lock
+                changeMutexLock(getClientPropertiesWithSystemMappingDisabled(), true);
+                return null;
+            }
+        });
+
+        hbaseTables = getHBaseTables();
+        assertTrue(hbaseTables.size() == PHOENIX_SYSTEM_TABLES.size());
+        assertTrue(hbaseTables.containsAll(PHOENIX_SYSTEM_TABLES));
+
+        try {
+            user3.doAs(new PrivilegedExceptionAction<Void>() {
+                @Override
+                public Void run() throws Exception {
+                    createConnection(getClientPropertiesWithSystemMappingEnabled());
```

Comment:
Can you verify you get a UpgradeInProgressException?

Reply:
Added this in the new commit.